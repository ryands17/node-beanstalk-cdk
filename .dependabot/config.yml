version: 1
update_configs:
  - package_manager: 'javascript'
    directory: '/'
    update_schedule: 'weekly'
    automerged_updates:
      - match:
          dependency_type: 'all'
          update_type: 'all'
    ignored_updates:
      - match:
          dependency_name: 'aws-cdk'
      - match:
          dependency_name: '@aws-cdk*'
    version_requirement_updates: 'widen_ranges'
