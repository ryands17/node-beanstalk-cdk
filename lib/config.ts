export const envVars = {
  REGION: process.env.REGION || 'us-east-1',
  APP_NAME: process.env.APP_NAME,
  REPO_OWNER: process.env.REPO_OWNER,
  REPO_NAME: process.env.REPO_NAME,
  APP_STAGE_NAME: process.env.APP_STAGE_NAME || 'develop',
  // change this to the branch of your choice
  BUILD_BRANCH: process.env.BUILD_BRANCH || '^refs/heads/main$',
};

export function validateEnvVariables() {
  for (let variable in envVars) {
    if (!envVars[variable as keyof typeof envVars])
      throw Error(`Environment variable ${variable} is not defined!`);
  }
}
