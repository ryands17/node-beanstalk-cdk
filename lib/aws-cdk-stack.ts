import * as cdk from '@aws-cdk/core';
import * as EB from '@aws-cdk/aws-elasticbeanstalk';
import * as IAM from '@aws-cdk/aws-iam';
import * as Codebuild from '@aws-cdk/aws-codebuild';
import * as cfg from './config';

export class AwsCdkStack extends cdk.Stack {
  constructor(scope: cdk.App, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // get platform to be created
    const platform = this.node.tryGetContext('platform');

    // beanstalk project setup
    const ebApp = new EB.CfnApplication(this, `${cfg.APP_NAME}-app`, {
      applicationName: cfg.APP_NAME,
    });

    // role for ec2 instance
    const options: EB.CfnEnvironment.OptionSettingProperty[] = [
      {
        namespace: 'aws:autoscaling:launchconfiguration',
        optionName: 'IamInstanceProfile',
        value: 'aws-elasticbeanstalk-ec2-role',
      },
    ];

    const ebEnv = new EB.CfnEnvironment(this, `${cfg.APP_NAME}-env`, {
      // default environmentName is `develop`
      environmentName: cfg.APP_STAGE_NAME,
      applicationName: ebApp.applicationName,
      platformArn: platform,
      optionSettings: options,
      solutionStackName:
        '64bit Amazon Linux 2 v0.1.0 running Node.js 10 (BETA)',
    });

    ebEnv.addDependsOn(ebApp);

    // codebuild project setup
    const webhooks: Codebuild.FilterGroup[] = [
      Codebuild.FilterGroup.inEventOf(
        Codebuild.EventAction.PUSH,
        Codebuild.EventAction.PULL_REQUEST_MERGED
      ).andHeadRefIs(cfg.BUILD_BRANCH),
    ];

    const repo = Codebuild.Source.gitHub({
      owner: cfg.REPO_OWNER,
      repo: cfg.REPO_NAME,
      webhook: true,
      webhookFilters: webhooks,
      reportBuildStatus: true,
    });

    const project = new Codebuild.Project(this, cfg.APP_NAME, {
      buildSpec: Codebuild.BuildSpec.fromSourceFilename('buildspec.yml'),
      projectName: `${cfg.APP_NAME}-build`,
      environment: {
        buildImage: Codebuild.LinuxBuildImage.AMAZON_LINUX_2_3,
        computeType: Codebuild.ComputeType.SMALL,
        environmentVariables: {
          EB_STAGE: {
            value: cfg.APP_STAGE_NAME,
          },
          // you can add more env variables here as per your requirement
        },
      },
      source: repo,
      timeout: cdk.Duration.minutes(20),
    });

    project.role.addManagedPolicy(
      IAM.ManagedPolicy.fromAwsManagedPolicyName(
        'AWSElasticBeanstalkFullAccess'
      )
    );
  }
}
