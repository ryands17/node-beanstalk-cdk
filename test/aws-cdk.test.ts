import { config } from 'dotenv';
config();
import * as cdk from '@aws-cdk/core';
import * as Codebuild from '@aws-cdk/aws-codebuild';
import {
  expect as expectCDK,
  haveResourceLike,
  SynthUtils,
} from '@aws-cdk/assert';
import { AwsCdkStack } from '../lib/aws-cdk-stack';
import { envVars } from '../lib/config';

test('Stack should synth to expected snaphsot', () => {
  const stack = createStack({ outdir: 'cdk.out', stackTraces: false });
  const snapshot = SynthUtils.synthesize(stack);
  expect(snapshot).toMatchSnapshot();
});

test('creates an Elastic Beanstalk app', () => {
  const stack = createStack();
  expectCDK(stack).to(
    haveResourceLike('AWS::ElasticBeanstalk::Application', {
      ApplicationName: 'node-eb',
    })
  );
});

test('creates an Elastic Beanstalk environment', () => {
  const stack = createStack();
  expectCDK(stack).to(
    haveResourceLike('AWS::ElasticBeanstalk::Environment', {
      ApplicationName: 'node-eb',
      EnvironmentName: 'main',
    })
  );
});

test('creates a Codebuild project', () => {
  const stack = createStack();
  expectCDK(stack).to(
    haveResourceLike('AWS::CodeBuild::Project', {
      Artifacts: {
        Type: 'NO_ARTIFACTS',
      },
      Environment: {
        ComputeType: Codebuild.ComputeType.SMALL,
        EnvironmentVariables: [
          {
            Name: 'EB_STAGE',
            Type: 'PLAINTEXT',
            Value: envVars.APP_STAGE_NAME,
          },
        ],
        PrivilegedMode: false,
      },
      Name: 'node-eb-build',
      TimeoutInMinutes: 20,
      Triggers: {
        FilterGroups: [
          [
            {
              Pattern: 'PUSH, PULL_REQUEST_MERGED',
              Type: 'EVENT',
            },
            {
              Pattern: envVars.BUILD_BRANCH,
              Type: 'HEAD_REF',
            },
          ],
        ],
        Webhook: true,
      },
    })
  );
});

function createStack(props?: cdk.AppProps) {
  const stackName = 'Node-EB-Deploy';
  const app = new cdk.App(props);
  return new AwsCdkStack(app, stackName);
}
