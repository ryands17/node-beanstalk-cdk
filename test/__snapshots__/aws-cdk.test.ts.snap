// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack should synth to expected snaphsot 1`] = `
CloudFormationStackArtifact {
  "_dependencyIDs": Array [],
  "_deps": Array [],
  "assembly": CloudAssembly {
    "artifacts": Array [
      [Circular],
      TreeCloudArtifact {
        "_dependencyIDs": Array [],
        "_deps": Array [],
        "assembly": [Circular],
        "file": "tree.json",
        "id": "Tree",
        "manifest": Object {
          "properties": Object {
            "file": "tree.json",
          },
          "type": "cdk:tree",
        },
        "messages": Array [],
      },
    ],
    "directory": "cdk.out",
    "manifest": Object {
      "artifacts": Object {
        "Node-EB-Deploy": Object {
          "environment": "aws://unknown-account/unknown-region",
          "metadata": Object {
            "/Node-EB-Deploy/node-eb-app": Array [
              Object {
                "data": "nodeebapp",
                "type": "aws:cdk:logicalId",
              },
            ],
            "/Node-EB-Deploy/node-eb-env": Array [
              Object {
                "data": "nodeebenv",
                "type": "aws:cdk:logicalId",
              },
            ],
            "/Node-EB-Deploy/node-eb/Resource": Array [
              Object {
                "data": "nodeeb21AA198B",
                "type": "aws:cdk:logicalId",
              },
            ],
            "/Node-EB-Deploy/node-eb/Role/DefaultPolicy/Resource": Array [
              Object {
                "data": "nodeebRoleDefaultPolicy56622678",
                "type": "aws:cdk:logicalId",
              },
            ],
            "/Node-EB-Deploy/node-eb/Role/Resource": Array [
              Object {
                "data": "nodeebRole83F66418",
                "type": "aws:cdk:logicalId",
              },
            ],
          },
          "properties": Object {
            "templateFile": "Node-EB-Deploy.template.json",
          },
          "type": "aws:cloudformation:stack",
        },
        "Tree": Object {
          "properties": Object {
            "file": "tree.json",
          },
          "type": "cdk:tree",
        },
      },
      "version": "9.0.0",
    },
    "runtime": Object {
      "libraries": Object {},
    },
    "version": "9.0.0",
  },
  "assets": Array [],
  "assumeRoleArn": undefined,
  "bootstrapStackVersionSsmParameter": undefined,
  "cloudFormationExecutionRoleArn": undefined,
  "displayName": "Node-EB-Deploy",
  "environment": Object {
    "account": "unknown-account",
    "name": "aws://unknown-account/unknown-region",
    "region": "unknown-region",
  },
  "id": "Node-EB-Deploy",
  "manifest": Object {
    "environment": "aws://unknown-account/unknown-region",
    "metadata": Object {
      "/Node-EB-Deploy/node-eb-app": Array [
        Object {
          "data": "nodeebapp",
          "type": "aws:cdk:logicalId",
        },
      ],
      "/Node-EB-Deploy/node-eb-env": Array [
        Object {
          "data": "nodeebenv",
          "type": "aws:cdk:logicalId",
        },
      ],
      "/Node-EB-Deploy/node-eb/Resource": Array [
        Object {
          "data": "nodeeb21AA198B",
          "type": "aws:cdk:logicalId",
        },
      ],
      "/Node-EB-Deploy/node-eb/Role/DefaultPolicy/Resource": Array [
        Object {
          "data": "nodeebRoleDefaultPolicy56622678",
          "type": "aws:cdk:logicalId",
        },
      ],
      "/Node-EB-Deploy/node-eb/Role/Resource": Array [
        Object {
          "data": "nodeebRole83F66418",
          "type": "aws:cdk:logicalId",
        },
      ],
    },
    "properties": Object {
      "templateFile": "Node-EB-Deploy.template.json",
    },
    "type": "aws:cloudformation:stack",
  },
  "messages": Array [],
  "name": "Node-EB-Deploy",
  "originalName": "Node-EB-Deploy",
  "parameters": Object {},
  "requiresBootstrapStackVersion": undefined,
  "stackName": "Node-EB-Deploy",
  "stackTemplateAssetObjectUrl": undefined,
  "tags": Object {},
  "templateFile": "Node-EB-Deploy.template.json",
  "terminationProtection": undefined,
}
`;
